File: Problems/0547. (M) Number of Provinces/Python3/solution.py
class Solution:
  def findCircleNum(self, isConnected: List[List[int]]) -> int:
    # iterative DFS

    n = len(isConnected)

    ret = 0
    visited = [False] * n
    for i in range(n):
      if not visited[i]:
        nodes = [i]
        while nodes:
          cur = nodes.pop()
          for nxt, c in enumerate(isConnected[cur]):
            if c and not visited[nxt]:
              nodes.append(nxt)
              visited[nxt] = True
        ret += 1
    
    return ret

File: Problems/0547. (M) Number of Provinces/Java/solution.java
class Solution {
  public int findCircleNum(int[][] isConnected) {
    // iterative DFS

    int n = isConnected.length;

    int ret = 0;
    boolean[] visited = new boolean[n];
    for (int i = 0; i < n; i++) {
      if (!visited[i]) {
        Deque<Integer> nodes = new ArrayDeque();
        nodes.push(i);
        while (nodes.size() > 0) {
          int cur = nodes.pop();
          for (int j = 0; j < n; j++) {
            if (isConnected[cur][j] == 1 && !visited[j]) {
              nodes.push(j);
              visited[j] = true;
            }
          }
        }
        ret++;
      }
    }
    return ret;
  }
}
